core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/KOptDescriptor.java

```diff
@@ -0,0 +1,271 @@
+                j = q[incl[p[i]]];
+                if (j >= i + 2 && (i & 1) == (j & 1) &&
+                        (s = (i & 1) == 1 ? Score(i + 1, j, k) :
+                                Score(i, j - 1, k)) > BestScore) {
+                    BestScore = s; Best_i = i; Best_j = j;
+                }
+            }
+            if (BestScore >= 0) {
+                i = Best_i; j = Best_j;
+                if ((i & 1) == 1) {
+                    out.add(FLIP(listVariableDescriptor, indexVariableSupply, entity,
+                                 t[p[i + 1]], t[p[i]], t[p[j]], t[p[j + 1]]));
+                    Reverse(i + 1, j);
+                } else {
+                    out.add(FLIP(listVariableDescriptor, indexVariableSupply, entity,
+                                 t[p[i - 1]], t[p[i]], t[p[j]], t[p[j - 1]]));
+                    Reverse(i, j - 1);
+                }
+                continue FindNextReversal;
+            }
+            for (i = 1; i <= 2 * k - 1; i += 2) {
+                j = q[incl[p[i]]];
+                if (j >= i + 2) {
+                    out.add(FLIP(listVariableDescriptor, indexVariableSupply, entity,
+                                 t[p[i]], t[p[i + 1]], t[p[j]], t[p[j - 1]]));
+                    Reverse(i + 1, j - 1);
+                    continue FindNextReversal;
+                }
+            }
+            isMoveNotDone = false;
+        }
+        return out;
+    }
+
+    public boolean isFeasible() {
+        int Count = 1, i = 2 * k;
+        while ((i = q[incl[p[i]]] ^ 1) != 0) {
+            Count++;
+        }
+        return (Count == k);
+    }
+
+    private void Reverse(int i, int j) {
+        for (; i < j; i++, j--) {
+            int pi = p[i];
+            q[p[i] = p[j]] = i;
+            q[p[j] = pi] = j;
+        }
+    }
+
+    private int Score(int left, int right, int k) {
+        int count = 0, i, j;
+        Reverse(left, right);
+        for (i = 1; i <= 2 * k - 2; i++) {
+            j = q[incl[p[i]]];
+            if (j >= i + 2 && (i & 1) == (j & 1)) {
+                count++;
+            }
+        }
+        Reverse(left, right);
+        return count;
+    }
+
+    private List2OptMove<Solution_> FLIP(ListVariableDescriptor<Solution_> listVariableDescriptor,
+                                         IndexVariableSupply indexVariableSupply,
+                                         Object entity,
+                                         Object firstEdgeStart,
+                                         Object firstEdgeEnd,
+                                         Object secondEdgeStart,
+                                         Object secondEdgeEnd) {
+        return new List2OptMove<>(listVariableDescriptor, indexVariableSupply, entity,
+                                  firstEdgeStart, firstEdgeEnd, secondEdgeEnd, secondEdgeStart);
+    }
+
+    public int getK() {
+        return k;
+    }
+
+    public String getMoveInfo() {
```

Comment:
This appears to be only used in tests. I suggest we move it there, and instead expose the data the methods needs. We certainly don't want to have to test this method.